File: test_operations.py
import unittest
import tempfile
import os
import yaml
import numpy as np
from YamlParser import InputConfigParser
from MeshObject import MeshObject
from QuadElement import Face, Cell
from FieldsHolder import FieldArray, FieldNames, DimType, MAX_DIM
from Operations import ComputeInteriorMassFlux, ComputeCellGradient, CFLTimeStepCompute
import math


class TestOperations(unittest.TestCase):
	def setUp(self):
		self.tmpdir = tempfile.mkdtemp()

	def tearDown(self):
		for f in os.listdir(self.tmpdir):
			os.remove(os.path.join(self.tmpdir, f))
		os.rmdir(self.tmpdir)
	def test_cfl_time_step_compute(self):
		cfg = {
			'mesh_parameters': {
				'x_range': [0, 2],
				'y_range': [0, 1],
				'num_cells_x': 4,
				'num_cells_y': 2
			}
		}
		path = os.path.join(self.tmpdir, 'mesh_cfl.yaml')
		with open(path, 'w') as f:
			yaml.safe_dump(cfg, f)

		parser = InputConfigParser(path)
		mesh = MeshObject(parser)
		cfl_value = 0.5
		time_step_op = CFLTimeStepCompute(mesh, cfl_value)

		dt = time_step_op.compute_time_step()
		expected_dt = 0.5 * min((2 - 0) / 4, (1 - 0) / 2)
		self.assertAlmostEqual(dt, expected_dt, places=12)
	def test_3x3_cell_grad_op_scalar(self):
		"""Test that we compute the right gradient given a linear field with slope 1"""
		cfg = {
			'mesh_parameters': {
				'x_range': [0, 1],
				'y_range': [0, 1],
				'num_cells_x': 3,
				'num_cells_y': 3
			}
		}
		path = os.path.join(self.tmpdir, 'mesh.yaml')
		with open(path, 'w') as f:
			yaml.safe_dump(cfg, f)

		parser = InputConfigParser(path)
		mesh = MeshObject(parser)
		mesh.generate_grid()
		pressureField = FieldArray(FieldNames.PRESSURE.value, DimType.SCALAR, mesh.get_num_cells())
		## initialize pressure field to be linear in x for testing gradient
		for i in range(mesh.get_num_cells()):
			pressureField.get_data()[i] = mesh.get_cells()[i].get_centroid()[0]  # Set pressure to x-coordinate of cell center
		gradPressureField = FieldArray(FieldNames.GRAD_PRESSURE.value, DimType.VECTOR, mesh.get_num_cells())
		gradient_op = ComputeCellGradient(mesh, pressureField, gradPressureField)
		gradient_op.compute_scalar_gradient()
        # for uniform grid, we expect the gradient to be approximately 1 in x direction and 0 in y direction in interior cells. Boundary cells will have different values due to one-sided difference...
		for interiorCell in mesh.get_interior_cells():
			cellIndex = interiorCell.get_flat_id()
			grad_x = gradPressureField.get_data()[2*cellIndex]  # x component of gradient
			grad_y = gradPressureField.get_data()[2*cellIndex + 1]  # y component of gradient
			self.assertAlmostEqual(grad_x, 1.0, places=5)
			self.assertAlmostEqual(grad_y, 0.0, places=5)
		

	def test_6x6_cell_grad_op_scalar_with_slope(self):
		"""Test that we compute the right gradient given a linear field, extended domain with a non-1 slope."""
		cfg = {
			'mesh_parameters': {
				'x_range': [0, 1],
				'y_range': [0, 1],
				'num_cells_x': 6,
				'num_cells_y': 6
			}
		}
		slope = 10.0
		path = os.path.join(self.tmpdir, 'mesh.yaml')
		with open(path, 'w') as f:
			yaml.safe_dump(cfg, f)

		parser = InputConfigParser(path)
		mesh = MeshObject(parser)
		mesh.generate_grid()
		pressureField = FieldArray(FieldNames.PRESSURE.value, DimType.SCALAR, mesh.get_num_cells())
		## initialize pressure field to be linear in x for testing gradient
		for i in range(mesh.get_num_cells()):
			pressureField.get_data()[i] = slope*mesh.get_cells()[i].get_centroid()[0]  # Set pressure to x-coordinate of cell center
		gradPressureField = FieldArray(FieldNames.GRAD_PRESSURE.value, DimType.VECTOR, mesh.get_num_cells())
		gradient_op = ComputeCellGradient(mesh, pressureField, gradPressureField)
		gradient_op.compute_scalar_gradient()
        # for uniform grid, we expect the gradient to be approximately 'slope' in x direction and 0 in y direction in interior cells. Boundary cells will have different values due to one-sided difference...
		for interiorCell in mesh.get_interior_cells():
			cellIndex = interiorCell.get_flat_id()
			grad_x = gradPressureField.get_data()[2*cellIndex]  # x component of gradient
			grad_y = gradPressureField.get_data()[2*cellIndex + 1]  # y component of gradient
			self.assertAlmostEqual(grad_x, slope, places=5)
			self.assertAlmostEqual(grad_y, 0.0, places=5)
	def test_3x3_cell_mass_flux_zero_pressure(self):
		"""Test that we get a summed mass flux of zero for a div-free velocity with zero pressure gradients."""
		numCellsX=3
		numCellsY=3
		cfg = {
			'mesh_parameters': {
				'x_range': [0, 1],
				'y_range': [0, 1],
				'num_cells_x': numCellsX,
				'num_cells_y': numCellsY
			}
		}
		path = os.path.join(self.tmpdir, 'mesh.yaml')
		with open(path, 'w') as f:
			yaml.safe_dump(cfg, f)

		parser = InputConfigParser(path)
		mesh = MeshObject(parser)
		mesh.generate_grid()
		pressureField = FieldArray(FieldNames.PRESSURE.value, DimType.SCALAR, mesh.get_num_cells())
		velocityField = FieldArray(FieldNames.VELOCITY_NEW.value, DimType.VECTOR, mesh.get_num_cells())
		## initialize pressure field to be linear in x for testing gradient
		for i in range(mesh.get_num_cells()):
			cell_x = mesh.get_cells()[i].get_centroid()[0]
			cell_y = mesh.get_cells()[i].get_centroid()[1]
			velocityField.get_data()[MAX_DIM*i] = math.pi*math.sin(math.pi*cell_x)*math.cos(math.pi*cell_y) 
			velocityField.get_data()[MAX_DIM*i + 1] = -math.pi*math.cos(math.pi*cell_x)*math.sin(math.pi*cell_y)
		pressureField.initialize_constant(0.0)
		gradPressureField = FieldArray(FieldNames.GRAD_PRESSURE.value, DimType.VECTOR, mesh.get_num_cells())
		gradient_op = ComputeCellGradient(mesh, pressureField, gradPressureField)
		gradient_op.compute_scalar_gradient()
		massFluxAlg = ComputeInteriorMassFlux(mesh, velocityField, pressureField, gradPressureField, dt=0.01)
		massFluxAlg.compute_mass_flux()
		faceListTest = np.zeros(mesh.get_num_cells())
		for face in mesh.get_internal_faces():
			leftCellID = face.get_left_cell()
			rightCellID = face.get_right_cell()
			faceListTest[leftCellID] += face.massFlux_
			faceListTest[rightCellID] -= face.massFlux_
		for cell in faceListTest:
			self.assertAlmostEqual(cell, 0.0, places=5)
	def test_3x3_cell_mass_flux_with_pressure(self):
		numCellsX=3
		numCellsY=3
		cfg = {
			'mesh_parameters': {
				'x_range': [0, 1],
				'y_range': [0, 1],
				'num_cells_x': numCellsX,
				'num_cells_y': numCellsY
			}
		}
		path = os.path.join(self.tmpdir, 'mesh.yaml')
		with open(path, 'w') as f:
			yaml.safe_dump(cfg, f)

		parser = InputConfigParser(path)
		mesh = MeshObject(parser)
		mesh.generate_grid()
		pressureField = FieldArray(FieldNames.PRESSURE.value, DimType.SCALAR, mesh.get_num_cells())
		velocityField = FieldArray(FieldNames.VELOCITY_NEW.value, DimType.VECTOR, mesh.get_num_cells())
		## initialize pressure field to be linear in x for testing gradient
		for i in range(mesh.get_num_cells()):
			cell_x = mesh.get_cells()[i].get_centroid()[0]
			cell_y = mesh.get_cells()[i].get_centroid()[1]
			pressureField.get_data()[i] = math.pi/2*(math.cos(2*math.pi*cell_x) + math.cos(2*math.pi*cell_y))
			velocityField.get_data()[MAX_DIM*i] = math.pi*math.sin(math.pi*cell_x)*math.cos(math.pi*cell_y) 
			velocityField.get_data()[MAX_DIM*i + 1] = -math.pi*math.cos(math.pi*cell_x)*math.sin(math.pi*cell_y)
		gradPressureField = FieldArray(FieldNames.GRAD_PRESSURE.value, DimType.VECTOR, mesh.get_num_cells())
		gradient_op = ComputeCellGradient(mesh, pressureField, gradPressureField)
		gradient_op.compute_scalar_gradient()
		massFluxAlg = ComputeInteriorMassFlux(mesh, velocityField, pressureField, gradPressureField, dt=0.01)
		massFluxAlg.compute_mass_flux()
		faceListTest = np.zeros(mesh.get_num_cells())
		for face in mesh.get_internal_faces():
			leftCellID = face.get_left_cell()
			rightCellID = face.get_right_cell()
			faceListTest[leftCellID] += face.massFlux_
			faceListTest[rightCellID] -= face.massFlux_

File: CPP_Impl/tests/test_operations.cpp
#include <petsc.h>
#include <gtest/gtest.h>
#include <fstream>
#include "../include/Operations.h"
#include "../include/MeshObject.h"
#include "../include/YamlParser.h"
#include "../include/FieldsHolder.h"
#include <vector>

using namespace Vinci4D;

class OperationsTest : public ::testing::Test {
protected:
    // Configurable parameters for test YAML
    int num_cells_x;
    int num_cells_y;
    double cfl;
    
    void SetUp() override {
        // Set default values (can be changed in individual tests before SetUp is called)
        num_cells_x = 20;
        num_cells_y = 20;
        cfl = 0.5;
        
        // Create test-specific YAML configuration with embedded parameters
        std::string yaml_content = "mesh_parameters:\n"
            "  x_range: [0.0, 1.0]\n"
            "  y_range: [0.0, 1.0]\n"
            "  num_cells_x: " + std::to_string(num_cells_x) + "\n"
            "  num_cells_y: " + std::to_string(num_cells_y) + "\n"
            "simulation:\n"
            "  Re: 100\n"
            "  CFL: " + std::to_string(cfl) + "\n";
        
        // Write to rank-specific temp file to avoid race conditions
        int rank = 0;
        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
        std::string configFile = "/tmp/operations_test_" + std::to_string(rank) + ".yaml";
        std::ofstream tmpYaml(configFile);
        tmpYaml << yaml_content;
        tmpYaml.close();
        
        parser = new InputConfigParser(configFile);
        mesh = new MeshObject(*parser);
        
        // For FieldArray sizing: track local and ghost cells separately
        int mpiSize = mesh->getMpiSize();
        if (mpiSize > 1) {
            numLocalCells = mesh->getNumLocalCells();
            numGhostCells = mesh->getNumGhostCells();
            // Synchronize all ranks at test start to prevent deadlocks in MPI operations
            MPI_Barrier(MPI_COMM_WORLD);
        } else {
            numLocalCells = mesh->getNumCells();
            numGhostCells = 0;
        }
        std::cout << " Finished setup...\n";
    }
    
    void TearDown() override {
        int mpiSize = mesh->getMpiSize();
        // Synchronize all ranks at test end to prevent deadlocks
        if (mpiSize > 1) {
            MPI_Barrier(MPI_COMM_WORLD);
        }
        delete mesh;
        delete parser;
    }
    
    InputConfigParser* parser;
    MeshObject* mesh;
    int numLocalCells;
    int numGhostCells;
};

TEST_F(OperationsTest, PerformanceTimer) {
    PerformanceTimer timer;
    
    EXPECT_NO_THROW(timer.startTimer("TestEvent"));
    
    // Do some work
    double sum = 0.0;
    for (int i = 0; i < 1000000; ++i) {
        sum += i;
    }
    
    EXPECT_NO_THROW(timer.endTimer());
    EXPECT_GT(sum, 0);
}

TEST_F(OperationsTest, CFLTimeStepCompute) {
    double CFL = parser->getCFL();
    CFLTimeStepCompute cflCompute(mesh, CFL);
    
    double dt = cflCompute.computeTimeStep();
    double dx = 1.0/num_cells_x;
    double expectedDt = cfl*dx/1.0;
    EXPECT_EQ(dt, expectedDt) << "Time step mismatch";
}

TEST_F(OperationsTest, ThreeByThreeCellGradOpScalar) {
    // Test that we compute the right gradient given a linear field with slope 1
    const char* yaml_content = R"(
mesh_parameters:
  x_range: [0.0, 1.0]
  y_range: [0.0, 1.0]
  num_cells_x: 20
  num_cells_y: 20
simulation:
  Re: 100
  CFL: 0.5
)";
    
    // Write temporary config (rank-specific to avoid race conditions)
    int rank = 0;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    std::string configFile = "/tmp/gradient_test_20x20_" + std::to_string(rank) + ".yaml";
    std::ofstream tmpYaml(configFile);
    tmpYaml << yaml_content;
    tmpYaml.close();
    
    InputConfigParser testParser(configFile);
    MeshObject testMesh(testParser);
    
    int mpiSize = testMesh.getMpiSize();
    int numLocal = (mpiSize > 1) ? testMesh.getNumLocalCells() : testMesh.getNumCells();
    int numGhost = (mpiSize > 1) ? testMesh.getNumGhostCells() : 0;
    
    // Initialize pressure field to be linear in x for testing gradient
    FieldArray pressureField("pressure", DimType::SCALAR, numLocal, numGhost);
    auto& pressureData = pressureField.getData();
    
    // Set field values in local+ghost cells
    const auto& cellsToInit = (mpiSize > 1) ? testMesh.getLocalCells() : testMesh.getCells();
    for (const auto& cell : cellsToInit) {
        int cellID = (mpiSize > 1) ? cell.getLocalId() : cell.getFlatId();
        auto centroid = cell.getCentroid();
        pressureData[cellID] = centroid[0];  // Set pressure to x-coordinate of cell center
    }
    
    // Compute gradient
    FieldArray gradPressureField("grad_pressure", DimType::VECTOR, numLocal, numGhost);
    
    // Exchange ghost cells before gradient computation in parallel mode
    if (mpiSize > 1) {
        pressureField.exchangeGhostCells(testMesh);
    }
    
    ComputeCellGradient gradientOp(&testMesh, &pressureField, &gradPressureField);
    gradientOp.computeScalarGradient();
    
    auto& gradData = gradPressureField.getData();
    
    // For uniform grid, expect gradient to be approximately 1 in x direction and 0 in y direction
    // in interior cells. Each rank checks only its local interior cells
    const auto& interiorCellsToCheck = (mpiSize > 1) ? 
        testMesh.getLocalInteriorCells() : testMesh.getInteriorCells();
    
    for (const auto* interiorCell : interiorCellsToCheck) {
        int cellID = (mpiSize > 1) ? interiorCell->getLocalId() : interiorCell->getFlatId();
        double grad_x = gradData[cellID * MAX_DIM];          // x component of gradient
        double grad_y = gradData[cellID * MAX_DIM + 1];      // y component of gradient
        
        EXPECT_NEAR(grad_x, 1.0, 1e-5) 
            << "X-gradient mismatch at interior cell (local ID: " << cellID << ")";
        EXPECT_NEAR(grad_y, 0.0, 1e-5) 
            << "Y-gradient mismatch at interior cell (local ID: " << cellID << ")";
    }
}

TEST_F(OperationsTest, TwentyByTwentyCellGradOpScalarWithSlope) {
    // Test that we compute the right gradient given a linear field with non-1 slope
    const char* yaml_content = R"(
mesh_parameters:
  x_range: [0.0, 1.0]
  y_range: [0.0, 1.0]
  num_cells_x: 20
  num_cells_y: 20
simulation:
  Re: 100
  CFL: 0.5
)";
    
    // Write temporary config (rank-specific to avoid race conditions)
    int rank = 0;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    std::string configFile = "/tmp/gradient_test_20x20_slope_" + std::to_string(rank) + ".yaml";
    std::ofstream tmpYaml(configFile);
    tmpYaml << yaml_content;
    tmpYaml.close();
    
    InputConfigParser testParser(configFile);
    MeshObject testMesh(testParser);
    
    int mpiSize = testMesh.getMpiSize();
    int numLocal = (mpiSize > 1) ? testMesh.getNumLocalCells() : testMesh.getNumCells();
    int numGhost = (mpiSize > 1) ? testMesh.getNumGhostCells() : 0;
    
    double slope = 10.0;
    
    // Initialize pressure field to be linear in x with given slope
    FieldArray pressureField("pressure", DimType::SCALAR, numLocal, numGhost);
    auto& pressureData = pressureField.getData();
    
    // Set field values in local+ghost cells
    const auto& cellsToInit = (mpiSize > 1) ? testMesh.getLocalCells() : testMesh.getCells();
    for (const auto& cell : cellsToInit) {
        int cellID = (mpiSize > 1) ? cell.getLocalId() : cell.getFlatId();
        auto centroid = cell.getCentroid();
        pressureData[cellID] = slope * centroid[0];  // Set pressure to slope * x-coordinate
    }
    
    // Compute gradient
    FieldArray gradPressureField("grad_pressure", DimType::VECTOR, numLocal, numGhost);
    
    // Exchange ghost cells before gradient computation in parallel mode
    if (mpiSize > 1) {
        pressureField.exchangeGhostCells(testMesh);
    }
    
    ComputeCellGradient gradientOp(&testMesh, &pressureField, &gradPressureField);
    gradientOp.computeScalarGradient();
    
    auto& gradData = gradPressureField.getData();
    
    // For uniform grid, expect gradient to be approximately 'slope' in x direction and 0 in y direction
    // in interior cells. Each rank checks only its local interior cells
    const auto& interiorCellsToCheck = (mpiSize > 1) ? 
        testMesh.getLocalInteriorCells() : testMesh.getInteriorCells();
    
    for (const auto* interiorCell : interiorCellsToCheck) {
        int cellID = (mpiSize > 1) ? interiorCell->getLocalId() : interiorCell->getFlatId();
        double grad_x = gradData[cellID * MAX_DIM];          // x component of gradient
        double grad_y = gradData[cellID * MAX_DIM + 1];      // y component of gradient
        
        EXPECT_NEAR(grad_x, slope, 1e-5) 
            << "X-gradient mismatch at interior cell " << cellID 
            << " (expected " << slope << ", got " << grad_x << ")";
        EXPECT_NEAR(grad_y, 0.0, 1e-5) 
            << "Y-gradient mismatch at interior cell " << cellID;
    }
    
    // Also check gradient at ghost cells (if running in parallel)
    gradPressureField.exchangeGhostCells(testMesh);
    if (mpiSize > 1) {
        const auto& ghostCells = testMesh.getGhostCells();
        for (size_t ghostIdx = 0; ghostIdx < ghostCells.size(); ++ghostIdx) {
            int ghostLocalId = numLocal + ghostIdx;
            double actual_grad_x = gradData[ghostLocalId * MAX_DIM];
            double actual_grad_y = gradData[ghostLocalId * MAX_DIM + 1];
            if (ghostCells[ghostIdx].getIndices()[0] == 0 || ghostCells[ghostIdx].getIndices()[0] == testMesh.getCellCount()[0] - 1 ||
                ghostCells[ghostIdx].getIndices()[1] == 0 || ghostCells[ghostIdx].getIndices()[1] == testMesh.getCellCount()[1] - 1) {
                    continue; //ignore true domain boundaries
            }
            EXPECT_NEAR(actual_grad_x, slope, 1e-4) 
                << "X-gradient mismatch at ghost cell (" << ghostCells[ghostIdx].getIndices()[0] << "," << ghostCells[ghostIdx].getIndices()[1] << ")"
                << " local ID = " << ghostLocalId;
            EXPECT_NEAR(actual_grad_y, 0.0, 1e-4) 
                << "Y-gradient mismatch at ghost cell (" << ghostCells[ghostIdx].getIndices()[0] << "," << ghostCells[ghostIdx].getIndices()[1] << ")"
                << " local ID = " << ghostLocalId;
        }
    }
}

int main(int argc, char** argv) {
    PetscInitialize(&argc, &argv, nullptr, nullptr);
    ::testing::InitGoogleTest(&argc, argv);
    
    // These tests now work in both sequential and parallel modes
    int result = RUN_ALL_TESTS();
    PetscFinalize();
    return result;
}
